coderd/provisionerjobs.go

```diff
@@ -334,98 +221,223 @@ func convertProvisionerJob(provisionerJob database.ProvisionerJob) codersdk.Prov
+			provisionersdk.ProvisionerJobLogsNotifyChannel(f.jobID),
+			f.listener,
+		)
+		if err != nil {
+			httpapi.Write(f.ctx, f.rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "failed to subscribe to job updates",
+				Detail:  err.Error(),
+			})
+			return
 		}
+		defer subCancel()
+
+		// we were provided `complete` prior to starting this subscription, so
+		// we also need to check whether the job is now complete, in case the
+		// job completed between the last time we queried the job and the start
+		// of the subscription.  If the job completes after this, we will get
+		// a notification on the subscription.
+		job, err := f.db.GetProvisionerJobByID(f.ctx, f.jobID)
+		if err != nil {
+			httpapi.Write(f.ctx, f.rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "failed to query job",
+				Detail:  err.Error(),
+			})
+			return
+		}
+		f.complete = jobIsComplete(f.logger, job)
 	}
 
-	closeSubscribe, err := api.Pubsub.Subscribe(
-		provisionerJobLogsChannel(jobID),
-		func(ctx context.Context, message []byte) {
-			if endOfLogs.Load() {
-				return
+	var err error
+	f.conn, err = websocket.Accept(f.rw, f.r, nil)
+	if err != nil {
+		httpapi.Write(f.ctx, f.rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Failed to accept websocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	defer f.conn.Close(websocket.StatusNormalClosure, "done")
+	go httpapi.Heartbeat(f.ctx, f.conn)
+
+	// query for logs once right away, so we can get historical data from before
+	// subscription
+	if err := f.query(); err != nil {
+		if f.ctx.Err() == nil && !xerrors.Is(err, io.EOF) {
+			// neither context expiry, nor EOF, close and log
+			f.logger.Error(f.ctx, "failed to query logs", slog.Error(err))
+			err = f.conn.Close(websocket.StatusInternalError, err.Error())
+			if err != nil {
+				f.logger.Warn(f.ctx, "failed to close webscoket", slog.Error(err))
 			}
-			jlMsg := provisionerJobLogsMessage{}
-			err := json.Unmarshal(message, &jlMsg)
+		}
+	}
+
+	// no need to wait if the job is done
+	if f.complete {
+		return
+	}
+	for {
+		select {
+		case err := <-f.errors:
+			// we've dropped at least one notification.  This can happen if we
+			// lose database connectivity.  We don't know whether the job is
+			// now complete since we could have missed the end of logs message.
+			// We could soldier on and retry, but loss of database connectivity
+			// is fairly serious, so instead just 500 and bail out.  Client
+			// can retry and hopefully find a healthier node.
+			f.logger.Error(f.ctx, "pubsub callback error", slog.Error(err))
+			err = f.conn.Close(websocket.StatusInternalError, err.Error())
 			if err != nil {
-				logger.Warn(ctx, "invalid provisioner job log on channel", slog.Error(err))
+				f.logger.Warn(f.ctx, "failed to close webscoket", slog.Error(err))
+			}
+			return
+		case <-f.ctx.Done():
+			// client disconnect
+			return
+		case n := <-f.kicks:
```

Comment:
I don't really understand the naming choice "kicks", why not "messages" or "notifications"?

Reply:
anyway, changing to notifications